{{cookiecutter.repo_name}}/devtools/scripts/initialize_conda.sh

```diff
@@ -0,0 +1,16 @@
+# $CI_OS is $matrix.os, as exported in GHA *.yaml
+# $CONDA (miniconda installation path) is always defined in the GHA virtual environments
+case ${CI_OS} in
+    windows*)
+        eval "$(${CONDA}/condabin/conda.bat shell.bash hook)";;
+    macOS*)
+        eval "$(${CONDA}/condabin/conda shell.bash hook)";;
+    *)
+        eval "$(conda shell.bash hook)";;
+esac
+
+if [[ -d ${CONDA}/envs/test ]] || [[ -d ${HOME}/.conda/envs/test ]]; then
```

Comment:
What if this took in args, so you could `devtools/scripts/conda.sh active test`?

Reply:
Good idea. I renamed the script so it's named `activate_conda_env.sh`, which will accept `$1`. If none is provided, it will run `conda activate`, but if you say `activate_conda_env.sh test` it will run `conda activate test`. Does that make sense?

Reply:
But why do that? If you allow the bash script take in arbitrary numbers of arguments then this could be used for `conda list` or other operations as well.

Reply:
Once you have called the script, you can execute `conda` normally:

```
. devtools/github-actions/activate_conda_env.sh
# base environment is now active
conda create -n test python=3.8 my_package
conda list
```

Would you like the activation step to be separate and explicit from the initialization? As in:

```
. devtools/github-actions/initialize_conda.sh
conda activate
```

Reply:
Bleh, I get that they do not particular like to keep state. But at the same time they are making it really hard to make a few operations easy. I vote a bit for `. devtools/github-actions/initialize_conda.sh` as it seems clear.

Reply:
One of the things I'd like to see implemented is defining custom shells per job, so you can do:

```
jobs:
  test:
    runs-on: {{ '${{ matrix.os }}' }}
    strategy:
      matrix:
        os: [macOS-latest, ubuntu-latest, windows-latest]
        python-version: [3.6, 3.7]
        bash-with-conda: ["bash --rcfile ~/devtools/github-actions/initialize_conda.sh"] 

[...]

    - name: Install testing dependencies
      shell: ${{ matrix.bash-with-conda }}
      run: |
        python -m pip install -U pytest pytest-cov codecov
```

This syntax is correct, but `shell` is not aware of the `${{ }}` expansions (yet?). For now we will need to resort to this script. I understand that we will have `initialize_conda.sh` only handling the pre-activation routines and then being explicit about further steps like `conda activate test`?

Reply:
Does `echo "::add-path::/path/to/dir"` help here?

Reply:
We _could_ add `$CONDA/condabin` to path to have the `conda` executables in `PATH`, but we still need to `eval` the bash hook and activate the environment anyway, so I don't see the need to pollute the path.

Reply:
If we added it to the path `conda init bash` should work across the board? Its mostly pedantic I know, but it looks much more similar to the operations you would complete on a local computer.

Reply:
Apparently the consensus is to use the bash thingy now: https://github.com/conda/conda/issues/7980#issuecomment-492784093. `conda init bash` modifies your `.bashrc` so this hook is automatically executed in every session. However, GHA's `bash` does not load profiles, and we would have to specify our own custom flags in the `shell` option anyway... I wish this could be simpler, but so far this (script) is the best option I've found.

Reply:
Is there a reason not to just use something like https://github.com/s-weigand/setup-conda to simplify this?

Reply:
It adds `$CONDA/bin` to `$PATH`, which is a bit too invasive depending on how you configure your `base` environment. This approach also requires the old `source activate <env>` mechanism, which is supposed to be deprecated (but still working). Anyway, it could stop working anytime.

The recommended way is using `eval` on the Bash hooks (as we are doing), per process. One cleaner workaround would be adding `$CONDA/condabin` (only contains `conda`) to `$PATH`, which we haven't tested, but again this would require `source activate`.

The best solution would be a mechanism in GHA to have each interpreter run the same command upon startup. If they didn't ignore `.bashrc` and friends, that is.

Reply:
I was more curious about something *like* that - i.e come up with a new one and release on Marketplace that fits the spec, or just modify the existing one, rather than advocating using *that* exact one.

Reply:
Oh, I see what you mean. The problem with releasing this as a GH Action is that we still need to execute it for each step because state is not kept across steps, and I don't think actions support that kind of behavior. I wonder if an Action can define a custom shell though; like "subclassing" their Bash shell with our own action and the using that... That'd be the cleanest, IMO.

Reply:
> still need to execute it for each step because state is not kept across steps
Ah perhaps this is a misunderstanding on my part - I though that state set by an action is then kept across all steps within the job, such that the environment should stay active throughout, but I could be wrong about this!

Reply:
Unfortunately each step is a separate Bash/shell process, so that kind of state is lost. Only thing you can keep across states is `$PATH` and custom environment variables, which is enough for `conda` to work with `source activate` but not the recommended approach (for some reason).